### ISSUE TITLE
Files Not Deleting From AWS S3 Bucket Storage

### ISSUE THREAD

**Issue by chris-mcdonald-dev** [directus/directus#18171]
### Describe the Bug

Similar issue to #7960 (a year old)

I'm fairly new to Directus and I've set up external storage with an AWS S3 bucket. When creating a collection and uploading an image file using the Directus admin panel, it uploads to my S3 bucket perfectly fine! However, when deleting a file, it deletes it from the Directus interface, but the files are left untouched in my AWS S3 bucket.

I assumed this was a permissions issue as well, but my S3 policy also includes `s3:deleteObject` (specifically only for the folder Directus is uploading to), and I'm using the key/secret from an IAM user with that policy assigned.

I've confirmed the permissions allow for deletion of the image files with AWS's SDK.  

Directus .env
```env
STORAGE_LOCATIONS="aws"

## AWS S3
STORAGE_AWS_ROOT="/folder"
STORAGE_AWS_DRIVER="s3"
STORAGE_AWS_BUCKET="bucket-name"
STORAGE_AWS_REGION="us-east-1"
STORAGE_AWS_ENDPOINT="s3.amazonaws.com"
STORAGE_AWS_KEY="iam-user-key"
STORAGE_AWS_SECRET="iam-user-secret"
```

AWS Policy Config
```json
{
  "Effect": "Allow",
  "Action": [
    "s3:PutObject",
    "s3:GetObject",
    "s3:DeleteObject"
  ],
  "Resource": "arn:aws:s3:::bucket-name/folder/*"
}
```

Any idea what I might be doing wrong?

### To Reproduce

1. Create an AWS S3 Bucket accessible with the policies listed below
2. Add credentials to Directus .env and create a new collection
3. Create a new item and upload an image (it should upload)
4. Try to delete the item/image (in my case it's remaining in my S3)

### Hosting Strategy

Self-Hosted (Custom)

**Comment by chris-mcdonald-dev** [comment]
**Update**: So, only slightly embarrassing 😅 I didn't realize you have to go to the file library to actually selete them from the file system.

It's still not deleting them from the S3 bucket, but it is giving an "Access Denied" error from AWS. So, progress lol ✨. So, it does seem like a permissions issue with my AWS policy.
I'm just not sure why the same exact credentials work for AWS's SDK when deleting files, but don't work with Directus.
I've also tried adding `s3:ListBucket`. Any idea what I need to add?

```json
{
  "Effect": "Allow",
  "Action": [
    "s3:PutObject",
    "s3:GetObject",
    "s3:ListBucket",
    "s3:DeleteObject"
  ],
  "Resource": "arn:aws:s3:::bucket-name/folder/*"
}
```

```console
ERROR: Access Denied
    err: {
      "type": "S3ServiceException",
      "message": "Access Denied",
      "stack":
          AccessDenied: Access Denied
      ....node_modules/@aws-sdk/middleware-signing/dist-cjs/middleware.js:14:20
}
```

**Comment by chris-mcdonald-dev** [comment]
# **Solution**

Found why it wasn't working!  
The way Directus interacts with AWS's S3 API is a bit different from AWS's JavaScript SDK when it comes to permissions (not sure if this is intended).

In addition to adding the "ListBucket" permission, in the resource property, you also need to add the actual bucket _separately_ in addition to the folder specified on the bucket.

### **AWS Policy**
```json
{
  "Effect": "Allow",
  "Action": [
    "s3:PutObject",
    "s3:GetObject",
    "s3:ListBucket",
    "s3:DeleteObject"
  ],
  "Resource": [
    "arn:aws:s3:::bucket-name/folder/*",
    "arn:aws:s3:::bucket-name"
  ]
}
```

Both the "ListBucket" as a permission and adding the actual bucket _in addition to_ adding the folder on the bucket as resources aren't things you need to do with the SDK. But they do appear to be required by Directus. It might be worth mentioning this somewhere in the docs if it's intentional. Thanks for everyone's troubleshooting on previous similar issues!

**Comment by rijkvanzanten** [comment]
> The way Directus interacts with AWS's S3 API is a bit different from AWS's JavaScript SDK when it comes to permissions (not sure if this is intended).

Mind sharing what you're comparing it to? 🙂  For what it's worth, Directus uses the `@aws-sdk/client-s3` as well, so it should be the same:

https://github.com/directus/directus/blob/c301cfbc82d1ce3ba8dad56800552c4899895b6e/packages/storage-driver-s3/src/index.ts#L160-L185

**Comment by chris-mcdonald-dev** [comment]
@rijkvanzanten That's what I thought was super weird too!

I tested the comparison initially with SDK in `aws-sdk`. Here's how I used it without the additional policy permissions. For whatever reason, it didn't give the same "Access Denied" error that I got when using it through Directus.

My simple example:
```js
const AWS = require('aws-sdk');

AWS.config.update({
  accessKeyId: "iam-user-key",
  secretAccessKey: "iam-user-secret"
});

const s3 = new AWS.S3();

const params = {
  Bucket: 'bucket-name',
  Key : "folder/file-name.png"
};

s3.deleteObject(params, (error) => {
  if (error) console.log(error);
});
```

I don't see anything drastically different in Directus' S3 client implementation. The only thought I have so far is that the `.deleteObject()` method used directly from the S3 instance in my example behaves a bit differently than sending a `DeleteObjectCommand` instance. 🤷🏾 Not sure, but at least there isn't too big of a difference in the solution!
